diff -r 2d8bc3c94ecb src/cmd/cc/funct.c
--- a/src/cmd/cc/funct.c	Fri Sep 21 17:10:44 2012 -0500
+++ b/src/cmd/cc/funct.c	Tue Apr 30 02:07:17 2013 -0400
@@ -269,7 +269,7 @@
 		goto bad;
 
 	f = alloc(sizeof(*f));
-	for(o=0; o<sizeof(f->sym); o++)
+	for(o=0; o<nelem(f->sym); o++)
 		f->sym[o] = S;
 
 	t->funct = f;
diff -r 2d8bc3c94ecb src/cmd/gc/bisonerrors
--- a/src/cmd/gc/bisonerrors	Fri Sep 21 17:10:44 2012 -0500
+++ b/src/cmd/gc/bisonerrors	Tue Apr 30 02:07:17 2013 -0400
@@ -27,7 +27,7 @@
 # In Grammar section of y.output,
 # record lhs and length of rhs for each rule.
 bison && /^Grammar/ { grammar = 1 }
-bison && /^(Terminals|state 0)/ { grammar = 0 }
+bison && /^(Terminals,|State 0)/ { grammar = 0 }
 grammar && NF>0 {
 	if($2 != "|") {
 		r = $2
@@ -41,9 +41,9 @@
 }
 
 # In state dumps, record shift/reduce actions.
-bison && /^state 0/ { grammar = 0; states = 1 }
+bison && /^State 0/ { grammar = 0; states = 1 }
 
-states && /^state / { state = $2 }
+states && /^State / { state = $2 }
 states { statetext[state] = statetext[state] $0 "\n" }
 
 states && / shift/ {
